File: backend/schemas/auto_scrape.py
from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class AutoScrapeStateUpdate(BaseModel):
    """Full state replacement (the SW pushes its full state object)."""

    state: dict[str, Any]


class HeartbeatRequest(BaseModel):
    extension_instance_id: Optional[str] = None


class WakeOrchestratorRequest(BaseModel):
    cycle_id: Optional[int] = None


class CycleCreate(BaseModel):
    started_at: datetime
    status: Literal["scrape_running"] = "scrape_running"


class CycleUpdate(BaseModel):
    status: Optional[
        Literal[
            "scrape_running",
            "scrape_complete",
            "postscrape_running",
            "post_scrape_complete",
            "failed",
        ]
    ] = None
    completed_at: Optional[datetime] = None
    phase_heartbeat_at: Optional[datetime] = None
    precheck_status: Optional[str] = None
    precheck_details: Optional[dict[str, Any]] = None
    scans_attempted: Optional[int] = None
    scans_succeeded: Optional[int] = None
    scans_failed: Optional[int] = None
    failures_by_reason: Optional[dict[str, int]] = None
    run_log_ids: Optional[list[UUID]] = None
    postcheck_status: Optional[str] = None
    postcheck_details: Optional[dict[str, Any]] = None
    cleanup_results: Optional[dict[str, Any]] = None
    dedup_task_id: Optional[UUID] = None
    match_results: Optional[dict[str, Any]] = None
    apply_results: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    notes: Optional[str] = None


class ConfigUpdate(BaseModel):
    enabled_sites: Optional[list[str]] = None
    keywords: Optional[list[str]] = None
    min_cycle_interval_minutes: Optional[int] = None
    inter_scan_delay_seconds: Optional[int] = None
    scan_timeout_minutes: Optional[int] = None
    max_consecutive_precheck_failures: Optional[int] = None
    max_consecutive_dead_session_cycles: Optional[int] = None
    run_dedup_after_scrape: Optional[bool] = None
    run_matching_after_dedup: Optional[bool] = None
    run_apply_after_matching: Optional[bool] = None


class SessionStateUpdate(BaseModel):
    last_probe_status: Literal["live", "expired", "captcha", "rate_limited", "unknown"]


class CleanupOrphanCyclesRequest(BaseModel):
    current_instance_id: str


class AutoScrapeStateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    state: dict[str, Any]
    last_sw_heartbeat_at: Optional[datetime]
    updated_at: datetime


class CycleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    cycle_id: int
    started_at: datetime
    completed_at: Optional[datetime]
    status: str
    phase_heartbeat_at: Optional[datetime]
    precheck_status: Optional[str]
    precheck_details: Optional[dict[str, Any]]
    scans_attempted: int
    scans_succeeded: int
    scans_failed: int
    failures_by_reason: Optional[dict[str, int]]
    run_log_ids: Optional[list[UUID]]
    postcheck_status: Optional[str]
    postcheck_details: Optional[dict[str, Any]]
    cleanup_results: Optional[dict[str, Any]]
    dedup_task_id: Optional[UUID]
    match_results: Optional[dict[str, Any]]
    apply_results: Optional[dict[str, Any]]
    error_message: Optional[str]
    notes: Optional[str]


class CycleCreateResponse(BaseModel):
    id: UUID
    cycle_id: int


class ConfigRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    config: dict[str, Any]
    updated_at: datetime


class ConfigUpdateResponse(BaseModel):
    config: dict[str, Any]
    warnings: list[str]
    next_cycle_estimated_at: Optional[datetime]


class ConfigLimitsResponse(BaseModel):
    limits: dict[str, dict[str, Any]]
    derived_limits: dict[str, Any]


class SiteSessionStateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    site: str
    last_probe_status: str
    last_probe_at: datetime
    consecutive_failures: int
    notified_user: bool
    backoff_multiplier: float
    updated_at: datetime


class CleanupInvalidEntriesResponse(BaseModel):
    deleted_jobs_empty_core: int
    deleted_jobs_empty_jd: int
    deleted_jobs_mismatched_website: int
    marked_failed_run_logs: int
    marked_failed_dedup_tasks: int

File: web/src/types/autoScrape.ts
import type { SourceSite } from '@/types/job'

/**
 * Mirrors the /admin/auto-scrape/* schemas (backend/schemas/auto_scrape.py).
 *
 * The old version of this file DID NOT COMPILE: it declared
 * `consecutive_precheck_failures` twice (line 18 required, line 24 optional)
 * -- a genuine TS2300 that shipped because no script ever ran tsc and ESLint
 * never touched .ts. Declared exactly once below.
 */

export interface AutoScrapeStateRead {
  /** Always 1 -- singleton, CHECK id=1. */
  id: number
  state: AutoScrapeState
  last_sw_heartbeat_at: string | null
  updated_at: string
}

/**
 * `state` is a FREE-FORM JSONB dict server-side -- not typed, not validated,
 * shaped only by a migration seed (023_auto_scrape_foundations.py:34-52).
 * This interface is OUR READING CONTRACT, not a server guarantee; hence the
 * index signature (Constitution Principle VII).
 */
export interface AutoScrapeState {
  enabled: boolean
  test_cycle_pending: boolean
  exit_requested: boolean
  config_change_pending: boolean
  cycle_id: number
  /** 'idle' | 'scrape_running' | 'postscrape_running'. Typed `string`, not a
   *  union: it is an opaque key in a free-form dict with no server validation
   *  (tests write 'test_put' into it). */
  cycle_phase: string
  extension_instance_id: string | null
  matrix_position: { site_index: number; keyword_index: number }
  cycle_results: {
    scans_attempted: number
    scans_succeeded: number
    scans_failed: number
    failures_by_reason: Record<string, number>
  }
  /** Declared ONCE. See the file header. */
  consecutive_precheck_failures: number
  /** Epoch MILLISECONDS. 0 | '0' | null all mean "unscheduled" -- the sentinel
   *  is polymorphic (routers/auto_scrape.py:88-98). Normalize on read. */
  next_cycle_at: number | string | null
  last_cycle_summary_id: string | null
  last_cycle_completed_at: string | null
  min_cycle_interval_ms: number
  clean_cycles_count: number
  [key: string]: unknown
}

export interface AutoScrapeInstances {
  instances: { instance_id: string; last_heartbeat_at: string }[]
  /** > 1 -> warn (FR-039). Concurrent instances corrupt cycle accounting. */
  count: number
}

/** NOTE the inconsistent underscore: `scrape_complete` but `post_scrape_complete`.
 *  As-built (schemas/auto_scrape.py:28-36). A real Literal server-side, unlike
 *  RunLog.status, so a closed union is honest here. */
export type CycleStatus =
  | 'scrape_running'
  | 'scrape_complete'
  | 'postscrape_running'
  | 'post_scrape_complete'
  | 'failed'

export interface Cycle {
  /** The uuid ROW id. */
  id: string
  /** The integer cycle NUMBER the operator sees (FR-041). Display this. */
  cycle_id: number
  started_at: string
  completed_at: string | null
  status: CycleStatus
  phase_heartbeat_at: string | null
  precheck_status: string | null
  precheck_details: Record<string, unknown> | null
  scans_attempted: number
  scans_succeeded: number
  scans_failed: number
  failures_by_reason: Record<string, number> | null
  run_log_ids: string[] | null
  postcheck_status: string | null
  postcheck_details: Record<string, unknown> | null
  cleanup_results: Record<string, unknown> | null
  /** ALWAYS null -- dedup retired. Not displayed (FR-004). */
  dedup_task_id: string | null
  match_results: Record<string, unknown> | null
  apply_results: Record<string, unknown> | null
  error_message: string | null
  notes: string | null
}

/** The one vocabulary with a real DB CHECK constraint
 *  (ck_site_session_states_probe_status, migration 023) as well as a Literal. */
export type ProbeStatus = 'live' | 'expired' | 'captcha' | 'rate_limited' | 'unknown'

export interface SiteSession {
  /** The PRIMARY KEY. There is no `id` field. */
  site: SourceSite
  last_probe_status: ProbeStatus
  last_probe_at: string
  consecutive_failures: number
  notified_user: boolean
  /** rate_limited doubles it, capped at 64.0. */
  backoff_multiplier: number
  updated_at: string
}

export interface AutoScrapeConfigRead {
  config: AutoScrapeConfig
  updated_at: string
}

export interface AutoScrapeConfig {
  enabled_sites: string[]
  keywords: string[]
  min_cycle_interval_minutes: number
  inter_scan_delay_seconds: number
  scan_timeout_minutes: number
  max_consecutive_precheck_failures: number
  max_consecutive_dead_session_cycles: number
  /** FR-045 dead fields: hidden, and NEVER sent (omission preserves them). */
  run_dedup_after_scrape: boolean
  run_matching_after_dedup: boolean
  run_apply_after_matching: boolean
}

export type AutoScrapeConfigUpdate = Partial<AutoScrapeConfig>

/** PUT /config returns THIS, while POST /config/reset returns the full
 *  AutoScrapeConfigRead envelope. The asymmetry is as-built. */
export interface AutoScrapeConfigUpdateResponse {
  config: AutoScrapeConfig
  /** sites x keywords >= 15 -> a warning on a 200, NOT an error. FR-044
   *  requires rendering these on success, not just field errors. */
  warnings: string[]
  next_cycle_estimated_at: string | null
}

export interface LimitRange {
  min: number
  max: number
  recommended: number
}

/** THE source of truth for orchestrator validation (FR-044). Never hardcode. */
export interface AutoScrapeConfigLimits {
  limits: Record<string, LimitRange>
  /** NOTE: valid_sites is nested INSIDE derived_limits in the response, even
   *  though get_limits() returns it as a sibling. Bind to the response. */
  derived_limits: {
    max_keywords: number
    max_scans_per_cycle_hard: number
    max_scans_per_cycle_warn: number
    valid_sites: string[]
  }
}

/** Derived, not a backend field. FR-038. */
export type HeartbeatGrade = 'fresh' | 'aging' | 'stale' | 'never'
